spark/src/main/java/org/apache/iceberg/RewriteManifestsAction.java

```diff
@@ -0,0 +1,490 @@
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.TypedColumn;
+import org.apache.spark.sql.expressions.Aggregator;
+import org.apache.spark.util.SerializableConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+// TODO: concurrent modification of snapshotIdInheritanceEnabled or specs?
+public class RewriteManifestsAction
+    implements SnapshotUpdateAction<RewriteManifestsAction, RewriteManifestsActionResult> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RewriteManifestsAction.class);
+
+  private final SparkSession spark;
+  private final JavaSparkContext sparkContext;
+  private final Table table;
+  private final FileIO fileIO;
+  private final Map<Integer, PartitionSpec> specs;
+  private final Map<String, String> summary;
+  private final int defaultParallelism;
+  private final boolean snapshotIdInheritanceEnabled;
+  private final long targetManifestSizeBytes;
+
+  private final Encoder<ManifestFile> manifestEncoder = Encoders.javaSerialization(ManifestFile.class);
+  private final Encoder<Entry> entryEncoder = Encoders.javaSerialization(Entry.class);
+  private final Encoder<Bin> binEncoder = Encoders.bean(Bin.class);
+
+  private Predicate<ManifestFile> predicate = manifest -> true;
+  private String stagingLocation = null;
+
+  RewriteManifestsAction(SparkSession spark, Table table) {
+    this.spark = spark;
+    this.sparkContext = new JavaSparkContext(spark.sparkContext());
+    this.table = table;
+    this.specs = table.specs();
+    this.summary = Maps.newHashMap();
+    this.defaultParallelism = Integer.parseInt(
+        spark.conf().get("spark.default.parallelism", "200"));
+    this.snapshotIdInheritanceEnabled = PropertyUtil.propertyAsBoolean(
+        table.properties(),
+        TableProperties.SNAPSHOT_ID_INHERITANCE_ENABLED,
+        TableProperties.SNAPSHOT_ID_INHERITANCE_ENABLED_DEFAULT);
+    this.targetManifestSizeBytes = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.MANIFEST_TARGET_SIZE_BYTES,
+        TableProperties.MANIFEST_TARGET_SIZE_BYTES_DEFAULT);
+
+    if (table.io() instanceof HadoopFileIO) {
+      // we need to use Spark's SerializableConfiguration to avoid issues with Kryo serialization
+      SerializableConfiguration conf = new SerializableConfiguration(((HadoopFileIO) table.io()).conf());
+      fileIO = new HadoopFileIO(conf::value);
+    } else {
+      fileIO = table.io();
+    }
+  }
+
+  public RewriteManifestsAction rewriteIf(Predicate<ManifestFile> newPredicate) {
+    this.predicate = newPredicate;
+    return this;
+  }
+
+  public RewriteManifestsAction stagingLocation(String newStagingLocation) {
+    this.stagingLocation = newStagingLocation;
+    return this;
+  }
+
+  @Override
+  public RewriteManifestsAction set(String property, String value) {
+    summary.put(property, value);
+    return this;
+  }
+
+  @Override
+  public RewriteManifestsActionResult execute() {
```

Comment:
I think it would simplify the implementation if you added a way to configure the spec (or spec id) you want to rewrite and default to the current spec.

I think it's a reasonable trade-off to process just one at a time. I can imagine a few cases where a table has multiple specs:
1. Transitioning from an old spec to a new one: we probably only need to process the old spec's files once since new data will be written with the new one.
2. When data is written using one spec and then transitioned to another for long-term storage. Like the first case, I think you'd only want to process one spec, the one that data is written with.

Focusing on a single spec allows you to simplify this.

Reply:
I agree processing one spec a time is a reasonable trade-off. Even if users need to rewrite all metadata, they can call this action for each spec. The latter can even be done in parallel.